### ISSUE TITLE
Question about model's input

### ISSUE THREAD

**Issue by Qiu-dot** [juho-lee/set_transformer#12]
Hi juho-lee,
I have many sets, each of which has a different size. I want to take some sets as a  mini-batch for set-transformer model. But I find that every set in a mini-batch must have same size. Have you ever face this problem? How did you deal with it? Padding or other methods?
thank you!

**Comment by juho-lee** [comment]
Hi,
You can actually process varying set sizes within a batch via masking attention (I have been using it personally so far).
The masking functionality is not included in the current code, but the implementation is pretty straightforward. 
You can instead use an official PyTorch implementation of multi-head attention which provides masking (https://pytorch.org/docs/stable/generated/torch.nn.MultiheadAttention.html).
Best wishes,
Juho

**Comment by Qiu-dot** [comment]
Thanks for your prompt reply!

**Comment by SRL94** [comment]
Hi,

I have the same question. If I have the input and I have padded the input to a matrix with the shape (batch_size, length, dim_input), how do I mask the input and send them to SetTransformer? Could you please explain more? Thank you very much.

Best regards,
Sirui

**Comment by Qiu-dot** [comment]
> Hi,
> 
> I have the same question. If I have the input and I have padded the input to a matrix with the shape (batch_size, length, dim_input), how do I mask the input and send them to SetTransformer? Could you please explain more? Thank you very much.
> 
> Best regards,
> Sirui

Hi Sirui,
You should pass the padding_mask position of your input to the forward function of nn.MultiHeadAttention. You can refer to the key_padding_mask of nn.MultiHeadAttention .(https://pytorch.org/docs/stable/generated/torch.nn.MultiheadAttention.html)
Best wishes,
Qiu

**Comment by SRL94** [comment]
Hi Qiu,

Thanks for the reply. But I did not find nn.MultiHeadAttention in the source code, how do I communicate nn.MultiHeadAttention with SetTransormer?

Best regards
Sirui

**Comment by Qiu-dot** [comment]
I think you need modify MAB of the source code when you use it.

**Comment by SRL94** [comment]
Thanks for reopening the issue, Qiu. Hope the developer will answer our questions and I will study the MAB as well :)

**Comment by mathDR** [comment]
I implemented the `MAB` module using the PyTorch multi-head attention.  My implementation looks like:
```
class MAB(nn.Module):
    def __init__(
        self,
        embed_dim: int,
        num_heads: int,
        key_padding_mask: Optional[torch.Tensor] = None,
        attn_mask: Optional[torch.Tensor] = None,
        ln: bool = False,
    ):
        super(MAB, self).__init__()
        self.embed_dim = embed_dim
        self.num_heads = num_heads
        self.key_padding_mask = key_padding_mask
        self.attn_mask = attn_mask
        self.multihead_attn = nn.MultiheadAttention(embed_dim, num_heads)

        if ln:
            self.ln0 = nn.LayerNorm(embed_dim)
            self.ln1 = nn.LayerNorm(embed_dim)

        self.fc_o = nn.Linear(embed_dim, embed_dim)

    def forward(self, X, Y):
        H = X + self.multihead_attn(
            X, Y, Y, key_padding_mask=self.key_padding_mask, attn_mask=self.attn_mask
        )
        H = H if getattr(self, 'ln0', None) is None else self.ln0(H)
        O = H + F.relu(self.fc_o(H))
        O = O if getattr(self, 'ln1', None) is None else self.ln1(O)
        return O
```
please comment if you see any errors or possible performance improvements

**Comment by mathDR** [comment]
although the mask should probably be passed into the `forward` function like so:  
```
class MAB(nn.Module):
    def __init__(self, embed_dim: int, num_heads: int, ln: bool = False):
        super(MAB, self).__init__()
        self.embed_dim = embed_dim
        self.num_heads = num_heads
        self.multihead_attn = nn.MultiheadAttention(embed_dim, num_heads)

        if ln:
            self.ln0 = nn.LayerNorm(embed_dim)
            self.ln1 = nn.LayerNorm(embed_dim)

        self.fc_o = nn.Linear(embed_dim, embed_dim)

    def forward(
        self,
        X,
        Y,
        key_padding_mask: Optional[torch.Tensor] = None,
        attn_mask: Optional[torch.Tensor] = None,
    ):
        H = X + self.multihead_attn(
            X, Y, Y, key_padding_mask=key_padding_mask, attn_mask=attn_mask
        )
        H = H if getattr(self, 'ln0', None) is None else self.ln0(H)
        O = H + F.relu(self.fc_o(H))
        O = O if getattr(self, 'ln1', None) is None else self.ln1(O)
        return O
```

**Comment by mathDR** [comment]
@juho-lee is the above about what you do within your scripts?

**Comment by juho-lee** [comment]
Yes, it is. Thanks.

**Comment by mathDR** [comment]
Okay great!  I appreciate the follow-up.  I would think this issue could be closed (as it solves the original problem being asked for the issue?)

**Comment by imskull** [comment]
I think there was a mistake, self-attention with mask is only permutation invariance at the very first layer! Eg. in the ISAB, the first MAB is PI but the second MAB is not PI anymore.

"It is worth noticing that masked self-attention is not permutation equivariant ([Tsai et al., 2019](https://www.sciencedirect.com/science/article/pii/S2666651022000146#b133))."

**Comment by Adithya1976** [comment]
If I only provide a key_padding_mask and not attn_mask there wouldn't be an issue with regards to permutation invariance right?
